design/one-pager-stacks-security-isolation.md

```diff
@@ -241,13 +277,209 @@ permissionScope: Cluster
+  admin, edit, view `ClusterRole` objects.
+ - Namespaced Stack: generate a set of `ClusterRole` objects that will be aggregated to the namespace
+  specific admin, edit, view `ClusterRole` objects.
+
+#### Default ClusterRoles
+
+The following built in roles will be installed and packaged with Crossplane, and updated with role aggregation as we
+ install additional stacks.
+
+The verbs supported in these roles(admin,edit,view) will model based on existing logic around
+ [built-in kubernetes roles](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#default-roles-and-role-bindings).
+
+|ClusterRole Name | Permissions |
+|:-------|:----------|
+| crossplane-admin | Admin: environment, namespaces, ClusterRole, Role, ClusterRoleBinding, RoleBinding |
+| crossplane-env-admin | Manage ClusterStacks, ... |
+| crossplane-env-edit | Edit Environment Resources, ... |
+| crossplane-env-view | View Environment Resources, Not permitted: secrets |
+
+We will also establish a convenience group - crossplane:masters - that binds to crossplane-admin with a
+ ClusterRoleBinding. This mirrors Kubernetes functionality for cluster-admin and system:master.
+
+Example running a command as crossplane-admin
+```
+kubectl {any crossplane-admin operation} --as-group=crossplane:masters
+```
+
+For discoverability of the roles for the environment we will add the following labels:
+ * `rbac.crossplane.io/role-environment: "true"`
+
+Match labels on these roles for auto-aggregation:
+ * `rbac.crossplane.io/aggregate-to-crossplane-admin: "true"`
+
+#### ClusterRoles for a given app namespace.
 
-With these two separate types, an administrator can allow access to installing cluster stacks to only certain restricted roles, then perhaps be more permissive with access to installing namespaced stacks.
+For any namespace that is annotated with `crossplane.io/managed-roles: true`, the stack-manager will install and manage
+ the following roles for that namespace that will be properly configured for ClusterRole aggregation as outlined below.
 
-If the install scope type does not match what the stack author has declared within their stack metadata, an error status should be set on the install resource and installation of the stack should not proceed.
+ * `crossplane:ns:{ns-name}:admin`
+ * `crossplane:ns:{ns-name}:edit`
+ * `crossplane:ns:{ns-name}:view`
+
+For discoverability of the roles for a given namespace we will add the following labels:
+ * `rbac.crossplane.io/role-namespace: "true"`
+ * `{namespace-name}: "true"`
+
+Match Roles:
+ * `rbac.crossplane.io/aggregate-to-{install-scope}-{role}: "true"`
+ * `{namespace-name}: "true" // to match specific namespace.`
+
+#### ClusterRoles for a given (Cluster)Stack Install
+
+For a stack given stack version it will have a unique set of roles in the format
+ {repo-name}:{stack-name}:{version}:{role=system|admin|edit|view}. System role is for internal use only and will be
```

Comment:
Where does `repo-name` come from?  A stack is currently aware of the app.yaml `Company`, and `Source` (registry.hub.docker.com, registry.crossplane.io)?  What should be used when source was omitted (deferring to local images and the default registry)?

Reply:
Special precaution will need to ensure that this is not set to `system`.

Would `stack` suffice as a prefix?

Reply:
Two stacks with the same name from different sources would collide on their cluster role if that were the case. hmm

Reply:
If the name is based on the mutable `Source`, we'll have to be careful about updating the roles to avoid orphaning them.

Reply:
Good idea on the prefix. Definitely needed. Any recommendation on what you would do with Source?

If a CRD that collides is installed from a different stack, is that a failure? 

Do we handle repo-name/stack-name as needing to be unique per crossplane, regardless of source, do we need to prefix everything with its source? Really open to ideas here.

Reply:
Luke pointed out in a later discussion and subsequent updates, that versioned stack ClusterRoles  would not be recommended for user rolebinding.  These clusterroles are only used for aggregation and for the the `system` role to provide the functionality needed for the controller.